### ISSUE TITLE
WayBar randomly crashes

### ISSUE THREAD

**Issue by GeoDerp** [Alexays/Waybar#320]
Hey guys, so I recently installed sway on my new Fedora 30 machine with waybar. And I have been experiencing quite a lot of crashes, I was wondering if this is a problem on my side or something els.

Thanks 
    [GeoDerp](https://github.com/GeoDerp)

**The common crash note:**
```
/usr/include/c++/9/bits/stl_stack.h:200: std::stack<_Tp, _Sequence>::reference std::stack<_Tp, _Sequence>::top() [with _Tp = Json::Value*; _Sequence = std::deque<Json::Value*, std::allocator<Json::Value*> >; std::stack<_Tp, _Sequence>::reference = Json::Value*&]: Assertion '__builtin_expect(!this->empty(), true)' failed.
Aborted (core dumped)
```

**Other crash notes:**
```
* Line 1, Column 21
  Missing ',' or ']' in array declaration

Workspaces: in Json::Value::resolveReference(key, end): requires objectValue
/usr/include/c++/9/bits/stl_stack.h:200: std::stack<_Tp, _Sequence>::reference std::stack<_Tp, _Sequence>::top() [with _Tp = Json::Value*; _Sequence = std::deque<Json::Value*, std::allocator<Json::Value*> >; std::stack<_Tp, _Sequence>::reference = Json::Value*&]: Assertion '__builtin_expect(!this->empty(), true)' failed.
Workspaces: * Line 1, Column 2
  Missing ':' after object member name
* Line 1, Column 24
  Syntax error: value, object or array expect
ed.
```
```
double free or corruption (out)
Aborted (core dumped)
```

**From gdb**:
```
Disabling module "backlight", No backlight found
Disabling module "battery", No batteries.
Disabling module "battery#bat2", No battery named BAT2
[New Thread 0x7fffaaffd700 (LWP 28544)]
[Thread 0x7fffe7b79700 (LWP 28514) exited]
/usr/include/c++/9/bits/stl_stack.h:200: std::stack<_Tp, _Sequence>::reference std::stack<_Tp, _Sequence>::top() [with _Tp = Json::Value*; _Sequence = std::deque<Json::Value*, std::allocator<Json::Value*> >; std::stack<_Tp, _Sequence>::reference = Json::Value*&]: Assertion '__builtin_expect(!this->empty(), true)' failed.

Thread 6 "waybar" received signal SIGABRT, Aborted.
[Switching to Thread 0x7fffe5fe1700 (LWP 28517)]
0x00007ffff6531eb5 in raise () from /lib64/libc.so.6
```

**WayBar version:**
```
Waybar v0.6.0
```

**My WayBar conf's:**
- [config](https://github.com/GeoDerp/SwayWM_Build/blob/master/dooooots/waybar/config) 
- [style.css](https://github.com/GeoDerp/SwayWM_Build/blob/master/dooooots/waybar/style.css)

**Comment by Alexays** [comment]
Theses crash must be fixed on master and in the upcoming version (0.6.4)

**Comment by Alexays** [comment]
```
* Line 1, Column 21
  Missing ',' or ']' in array declaration
```
Is a syntax pb from your config file

**Comment by GeoDerp** [comment]
> ```
> * Line 1, Column 21
>   Missing ',' or ']' in array declaration
> ```
> 
> Is a syntax pb from your config file

I don't believe so, After a quick check I could't find anything. Do Keep in mind that:
 ```shell
 * Line 1, Column 21
  Missing ',' or ']' in array declaration
```
Does not come up often or at all comparing to: 
```shell
/usr/include/c++/9/bits/stl_stack.h:200: std::stack<_Tp, _Sequence>::reference std::stack<_Tp, _Sequence>::top() [with _Tp = Json::Value*; _Sequence = std::deque<Json::Value*, std::allocator<Json::Value*> >; std::stack<_Tp, _Sequence>::reference = Json::Value*&]: Assertion '__builtin_expect(!this->empty(), true)' failed.
Aborted (core dumped)
```
which can happen from every 2 min's to every 30min's

**Comment by GeoDerp** [comment]
![config](https://user-images.githubusercontent.com/18461782/57858856-f5da5280-7830-11e9-92ab-938a7588e1e6.jpg)

Probably can't see it, but there isn't too much  syntax wise that makes it different from the master (master is the one on the left)

**Comment by GeoDerp** [comment]
I have also gotten this just then:
```
Workspaces: assert json failed
* Line 1, Column 21
  Missing ',' or ']' in array declaration

Workspaces: assert json failed
```
This wasn't a crash error, but an error none then less

**Comment by Alexays** [comment]
No i mean by the master version, waybar itself, not configs files :)

**Comment by GeoDerp** [comment]
> No i mean by the master version, waybar itself, not configs files :)

hahaha, aaaaaaah, sorry it's been a long day. I'm currently using the 0.6.3 tag but ill compile the master into a rpm and try that

**Comment by firat** [comment]
Built Waybar from master ([a1ffa7f](https://github.com/Alexays/Waybar/tree/a1ffa7fa9f271dc2af1bc9d8fbf2e280365a312f)) today. I still get seemingly random crashes. Built and ran Waybar with debugoptimized and got the following BT. Hope it helps.

https://gist.github.com/firat/8653f6c63596383d8eb703a62d9ae4d4

**Comment by Alexays** [comment]
@firat It doesn't help me much, do you do anything in particular before the crash?
Does this happen often?
If this happens often can you test by removing modules one by one to see if it comes from a particular module? :)

**Comment by firat** [comment]
@Alexays it's hard to tell. I usually takes me a while to notice that Waybar is gone. It's hard to notice it crashed. Maybe sending a notification when it crashes might be a solution. I'll see if I can write a wrapper script to run waybar and send a notification when it exits.

**Comment by firat** [comment]
This seems to be happening when I plug/unplug an external display.

**Comment by fmartingr** [comment]
This happens to me randomly too with latest git. Best I can say is when switching workspaces. How can I run `waybar` to get more debugging information?

**Comment by Alexays** [comment]
Okay, I figured it out this came from the network module.
Can you test if it's better with d34c3a801c3ab73eb70b02dc0bea25b4ae1bb81d

**Comment by fmartingr** [comment]
I updated from master (because I read the email notification directly) and now some icons are missing, but I'm sure you're aware of it because the edits. :+1:

**Comment by Alexays** [comment]
Some icons?

**Comment by fmartingr** [comment]
> Some icons?

Yeah, after updating from master by ethernet/vpn icons are gone:

![image](https://user-images.githubusercontent.com/812088/57969177-29db8200-79af-11e9-8f54-b3f97df97402.png)

![image](https://user-images.githubusercontent.com/812088/57969188-50012200-79af-11e9-89a0-0101cc98c585.png)

**Comment by Alexays** [comment]
Oh i see, fixed on master

**Comment by fmartingr** [comment]
> Oh i see, fixed on master

Confirmed :+1: Thanks! (and sorry for the offtopic on the issue)

**Comment by Alexays** [comment]
Fixed (on my side) with aa385e28b6c2c6e8e261b6b108f3635e86796b01
I'm not a big fan of the fix but it works.
Tell me if it's not fixed for you.

**Comment by fmartingr** [comment]
Mine still crashes randomly, though a lot less than before:
<details>
<summary>strace</summary>

```
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=13, revents=POLLIN}])
read(13, "\270\320EE\240U\0\0\220LHE\240U\0\0", 16) = 16
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=13, revents=POLLIN}])
read(13, "\270\320EE\240U\0\0\220LHE\240U\0\0", 16) = 16
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=13, revents=POLLIN}])
read(13, "(\232HE\240U\0\0\220LHE\240U\0\0", 16) = 16
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
sendmsg(4, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\25\0\0\0\1\0\24\0\35\0\0\0\0\0\0\0\0\0\0\0\25\0\0\0\10\0\f\0\1\0\0\0"..., iov_len=76}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, MSG_DONTWAIT|MSG_NOSIGNAL) = 76
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=4, revents=POLLIN}])
recvmsg(4, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\35\0\0\0\0\0\10\0", iov_len=3452}, {iov_base="", iov_len=644}], msg_iovlen=2, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 8
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=4, revents=POLLIN}])
recvmsg(4, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\"\0\0\0\0\0\f\0%\341\374\0\1\0\0\0\1\0\f\0\"\0\0\0", iov_len=3444}, {iov_base="", iov_len=652}], msg_iovlen=2, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 24
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=13, revents=POLLIN}])
read(13, "\270\320EE\240U\0\0\220LHE\240U\0\0", 16) = 16
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=13, revents=POLLIN}])
read(13, "\270\320EE\240U\0\0\220LHE\240U\0\0", 16) = 16
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=13, revents=POLLIN}])
read(13, "\270\320EE\240U\0\0\220LHE\240U\0\0", 16) = 16
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 2 ([{fd=6, revents=POLLIN}, {fd=13, revents=POLLIN}])
read(6, "\1\0\0\0\0\0\0\0", 16)         = 8
read(13, "\10\324NE\240U\0\0\220LHE\240U\0\0", 16) = 16
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 15) = 0 (Timeout)
sendmsg(4, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\25\0\0\0\1\0\24\0\35\0\0\0\0\0\0\0\0\0\0\0\25\0\0\0\10\0\f\0\1\0\0\0"..., iov_len=76}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, MSG_DONTWAIT|MSG_NOSIGNAL) = 76
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=4, revents=POLLIN}])
recvmsg(4, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\35\0\0\0\0\0\10\0", iov_len=3420}, {iov_base="", iov_len=676}], msg_iovlen=2, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 8
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=4, revents=POLLIN}])
recvmsg(4, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\"\0\0\0\0\0\f\0\5\v\375\0\1\0\0\0\1\0\f\0\"\0\0\0", iov_len=3412}, {iov_base="", iov_len=684}], msg_iovlen=2, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 24
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=13, revents=POLLIN}])
read(13, "X\237GE\240U\0\0\220LHE\240U\0\0", 16) = 16
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
sendmsg(4, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\25\0\0\0\1\0\24\0\35\0\0\0\0\0\0\0\0\0\0\0\25\0\0\0\10\0\f\0\1\0\0\0"..., iov_len=76}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, MSG_DONTWAIT|MSG_NOSIGNAL) = 76
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=4, revents=POLLIN}])
recvmsg(4, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\35\0\0\0\0\0\10\0", iov_len=3388}, {iov_base="", iov_len=708}], msg_iovlen=2, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 8
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=4, revents=POLLIN}])
recvmsg(4, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\"\0\0\0\0\0\f\0\234\21\375\0\1\0\0\0\1\0\f\0\"\0\0\0", iov_len=3380}, {iov_base="", iov_len=716}], msg_iovlen=2, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 24
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 2 ([{fd=6, revents=POLLIN}, {fd=13, revents=POLLIN}])
read(6, "\1\0\0\0\0\0\0\0", 16)         = 8
read(13, "\10\324NE\240U\0\0\220LHE\240U\0\0", 16) = 16
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 14) = 0 (Timeout)
sendmsg(4, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\25\0\0\0\1\0\24\0\35\0\0\0\0\0\0\0\0\0\0\0\25\0\0\0\10\0\f\0\1\0\0\0"..., iov_len=76}], msg_iovlen=1, msg_controllen=0, msg_flags=0}, MSG_DONTWAIT|MSG_NOSIGNAL) = 76
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=4, revents=POLLIN}])
recvmsg(4, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\35\0\0\0\0\0\10\0", iov_len=3356}, {iov_base="", iov_len=740}], msg_iovlen=2, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 8
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=4, revents=POLLIN}])
recvmsg(4, {msg_name=NULL, msg_namelen=0, msg_iov=[{iov_base="\"\0\0\0\0\0\f\0z\24\375\0\1\0\0\0\1\0\f\0\"\0\0\0", iov_len=3348}, {iov_base="", iov_len=748}], msg_iovlen=2, msg_controllen=0, msg_flags=MSG_CMSG_CLOEXEC}, MSG_DONTWAIT|MSG_CMSG_CLOEXEC) = 24
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=13, revents=POLLIN}])
read(13, "\270\320EE\240U\0\0\220LHE\240U\0\0", 16) = 16
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
lseek(3, 0, SEEK_SET)                   = 0
read(3, "TcpExt: SyncookiesSent Syncookie"..., 8191) = 2588
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=6, revents=POLLIN}])
read(6, "\1\0\0\0\0\0\0\0", 16)         = 8
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, -1) = 1 ([{fd=13, revents=POLLIN}])
read(13, "\10\324NE\240U\0\0\220LHE\240U\0\0", 16) = 16
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 0) = 0 (Timeout)
poll([{fd=4, events=POLLIN}, {fd=6, events=POLLIN}, {fd=11, events=POLLIN}, {fd=13, events=POLLIN}], 4, 15) = 0 (Timeout)
--- SIGSEGV {si_signo=SIGSEGV, si_code=SEGV_MAPERR, si_addr=0x7ffd00000000} ---
+++ killed by SIGSEGV (core dumped) +++
```
</details>
